homeassistant/setup.py

```diff
@@ -142,23 +142,37 @@ async def async_setup_component(
     setup_futures: dict[str, asyncio.Future[bool]] = hass.data.setdefault(
         DATA_SETUP, {}
     )
+    setup_done_futures: dict[str, asyncio.Future[bool]] = hass.data.setdefault(
+        DATA_SETUP_DONE, {}
+    )
 
-    if existing_future := setup_futures.get(domain):
-        return await existing_future
+    if existing_setup_future := setup_futures.get(domain):
+        return await existing_setup_future
 
-    future = hass.loop.create_future()
-    setup_futures[domain] = future
+    setup_future = hass.loop.create_future()
+    setup_futures[domain] = setup_future
 
     try:
         result = await _async_setup_component(hass, domain, config)
-        future.set_result(result)
+        setup_future.set_result(result)
+        if setup_done_future := setup_done_futures.pop(domain, None):
+            setup_done_future.set_result(result)
         return result
     except BaseException as err:  # pylint: disable=broad-except
-        future.set_exception(err)
+        setup_future.set_exception(err)
+        with contextlib.suppress(BaseException):
+            # Clear the flag as its normal that nothing
+            # will wait for this future to be resolved
+            # if there are no concurrent setup attempts
+            await setup_future
+        if setup_done_future := setup_done_futures.pop(domain, None):
+            setup_done_future.set_exception(err)
+            with contextlib.suppress(BaseException):
+                # Clear the flag as its normal that nothing
+                # will wait for this future to be resolved
+                # if there are no concurrent setup attempts
+                await setup_done_future
```

Comment:
I really want to find a way to get rid of `setup` and `setup_done` so we don't have to have both